meteor/server/cronjobs.ts

```diff
@@ -70,59 +79,101 @@ async function cleanupOldDataCronjob() {
+		},
+		{
+			projection: {
+				_id: 1,
+				subType: 1,
+				parentDeviceId: 1,
+				lastSeen: 1,
+			},
+		}
+	)) as Array<Pick<PeripheralDevice, '_id' | 'subType' | 'parentDeviceId' | 'lastSeen'>>
 
-		for (const device of peripheralDevices) {
-			const subDevices = await PeripheralDevices.findFetchAsync({
-				parentDeviceId: device._id,
-			})
+	const deviceMap = normalizeArrayToMap(casparcgAndParentDevices, '_id')
 
-			for (const subDevice of subDevices) {
-				if (subDevice.type === PeripheralDeviceType.PLAYOUT && subDevice.subType === TSR.DeviceType.CASPARCG) {
-					logger.info('Cronjob: Trying to restart CasparCG on device "' + subDevice._id + '"')
-
-					ps.push(
-						executePeripheralDeviceFunctionWithCustomTimeout(
-							subDevice._id,
-							DEFAULT_TSR_ACTION_TIMEOUT_TIME,
-							{
-								actionId: TSR.CasparCGActions.RestartServer,
-								payload: {},
-							}
-						)
-							.then(() => {
-								logger.info('Cronjob: "' + subDevice._id + '": CasparCG restart done')
-							})
-							.catch((err) => {
-								logger.error(
-									`Cronjob: "${subDevice._id}": CasparCG restart error: ${stringifyError(err)}`
-								)
-
-								if ((err + '').match(/timeout/i)) {
-									// If it was a timeout, maybe we could try again later?
-									if (failedRetries < 5) {
-										failedRetries++
-										lastNightlyCronjob = previousLastNightlyCronjob // try again later
-									}
-								} else {
-									// Propogate the error
-									throw err
-								}
-							})
-					)
-				}
-			}
+	for (const device of casparcgAndParentDevices) {
+		if (device.subType !== TSR.DeviceType.CASPARCG) continue
+
+		if (device.lastSeen < Date.now() - CASPARCG_LAST_SEEN_PERIOD_MS) {
+			logger.info(`Cronjob: Skipping CasparCG device "${device._id}" offline`)
+			shouldRetryAttempt = true
+			continue
+		}
+
+		if (!device.parentDeviceId) {
+			logger.info(`Cronjob: Skipping CasparCG device "${device._id}" without parentDeviceId`)
+			// Misconfiguration, don't retry
+			continue
+		}
+		const parentDevice = deviceMap.get(device.parentDeviceId)
+		if (!parentDevice) {
+			logger.info(`Cronjob: Skipping CasparCG device "${device._id}" without missing parent device`)
+			// Misconfiguration, don't retry
+			continue
+		}
+
+		if (parentDevice.lastSeen < Date.now() - CASPARCG_LAST_SEEN_PERIOD_MS) {
+			logger.info(`Cronjob: Skipping CasparCG device "${device._id}" with offline parent device`)
+			shouldRetryAttempt = true
+			continue
 		}
+
+		logger.info(`Cronjob: Trying to restart CasparCG on device "${device._id}"`)
+
+		ps.push(
+			executePeripheralDeviceFunctionWithCustomTimeout(device._id, DEFAULT_TSR_ACTION_TIMEOUT_TIME, {
+				actionId: TSR.CasparCGActions.RestartServer,
+				payload: {},
+			})
+				.then((res) => {
+					if (res.result === TSR.ActionExecutionResultCode.Ok) {
+						logger.info(`Cronjob: "${device._id}": CasparCG restart done`)
```

Comment:
Shouldn't a succesful restart of one device be recorded, so that it can be skipped if another device fails and `shouldRetryAttempt` be true? Otherwise, won't be restarting all CasparCG multiple times, even though only one of them fails to restart?

Reply:
Yes, I agree that it should be more intelligent. But I would class that as being more of a feature than a fix, so I wrote up a new task to improve that SOFIE-2684. I think it will end up being a bit fiddly, so am not comfortable doing that as part of a fix for r50.